library/std/src/sync/reentrant_lock.rs

```diff
@@ -303,18 +385,23 @@ impl<T: ?Sized> Drop for ReentrantLockGuard<'_, T> {
         unsafe {
             *self.lock.lock_count.get() -= 1;
             if *self.lock.lock_count.get() == 0 {
-                self.lock.owner.store(0, Relaxed);
+                self.lock.owner.set(0);
                 self.lock.mutex.unlock();
             }
         }
     }
 }
 
-/// Get an address that is unique per running thread.
+/// Returns the current thread's ThreadId value, which is guaranteed
+/// to be unique across the lifetime of the process.
 ///
-/// This can be used as a non-null usize-sized ID.
-pub(crate) fn current_thread_unique_ptr() -> usize {
-    // Use a non-drop type to make sure it's still available during thread destruction.
-    thread_local! { static X: u8 = const { 0 } }
-    X.with(|x| <*const _>::addr(x))
+/// Panics if called during a TLS destructor on a thread that hasn't
+/// been assigned an ID.
+pub(crate) fn current_thread_id() -> u64 {
+    #[cold]
+    fn no_tid() -> ! {
+        panic!("Thread hasn't been assigned an ID!")
```

Comment:
This should probably be `rtabort!` or so, `panic!` may call user code which may try to print things which may try to do reentrant lock stuff again.

Reply:
Hm, is there a meaningful difference between this and something like `RwLock::read` (which panics when the reader count hits the max value)? In both cases, if you try to relock it then it'll just panic again.

Reply:
> IIRC ReentrantLock is used deep in the guts of std, unlike RwLock? 

That makes sense, I'll change it to an `rtabort!` then.

> Other parts of your PR even assume for correctness that such overflow never happens.

Are you referring to the `seq` counter? That's the only place I can think of where overflow can matter and that can only cause problems in the exact case described in the PR. There's also precedence for this, e.g. `Condvar` also operates like this (though you could argue that the fallout from hitting this case in `Condvar` is smaller as it just keeps a thread asleep indefinitely instead of potentially causing unsoundness).